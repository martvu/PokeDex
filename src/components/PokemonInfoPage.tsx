import { usePokemonData, useSpeciesData } from '../utils/pokeApi';
import { useState } from 'react';
import { Link, useParams } from 'react-router-dom';
import {
  TYPE_COLORS,
  STAT_COLORS,
  MAX_NO_OF_POKEMON,
} from '../utils/constants';
import { FlavorText, PokemonStat, PokemonType } from '../utils/pokeApiTypes';
import { formatPokemonName, getTypeColorGradient } from '../utils/utils';
import Header from './Header';

export default function PokemonInfoPage() {
  const { id } = useParams();
  const [currentId, setCurrentId] = useState<number>(id ? parseInt(id) : 1);
  const {
    data: pokemonDetails,
    isLoading,
    error,
  } = usePokemonData(currentId.toString());
  const {
    data: speciesData,
    isLoading: isLoadingSpecies,
    error: speciesError,
  } = useSpeciesData(currentId.toString());

  let gradient = 'black';
  if (pokemonDetails) {
    gradient = getTypeColorGradient(pokemonDetails);
  }

  return (
    <>
      <Header />
      <div className="filler-div"></div>
      <div className="info-page-container">
        <div className="info-home-btn"></div>
        {isLoading || isLoadingSpecies && <div>Loading...</div>}
        {error || speciesError && <div>Error fetching data</div>}
        {pokemonDetails && speciesData && (
          <>
            <div className="info-pokemon-card" style={{ background: gradient }}>
              <div className="pokemon-name-id">
                <h1>{formatPokemonName(pokemonDetails?.name)}</h1>
                <h4 className="pokemon-text">
                  {'#' + ('00' + pokemonDetails?.id).slice(-3)}
                </h4>
              </div>
              <div className="info-pokemon-img">
                <img
                  className="info-pokemon-sprite"
                  src={
                    pokemonDetails?.sprites?.other?.['official-artwork']
                      ?.front_default || '#'
                  }
                  alt={pokemonDetails.name}
                />
              </div>
              <div className="info-pokemon-data">
                <div className="info-type-list">
                  {pokemonDetails.types.map((type: PokemonType) => (
                    <span
                      key={type.slot}
                      className="info-type-badge"
                      style={{
                        backgroundColor: `${TYPE_COLORS[type.type.name]}`,
                      }}
                    >
                      {type.type.name.charAt(0).toUpperCase() +
                        type.type.name.slice(1)}
                    </span>
                  ))}
                </div>
                <div className="pokemon-genus">
                  {isLoadingSpecies
                    ? 'Loading...'
                    : speciesData?.genera[7]?.genus}
                </div>
              </div>
            </div>
            <div>
              <Link to={`/project1/pokemon/${(currentId - 1).toString()}`}>
                <button
                  onClick={() => setCurrentId(currentId - 1)}
                  disabled={currentId === 1}
                  className="prev-btn"
                >
                  Prev
                </button>
              </Link>
              <Link to={`/project1/pokemon/${(currentId + 1).toString()}`}>
                <button
                  onClick={() => setCurrentId(currentId + 1)}
                  disabled={currentId === MAX_NO_OF_POKEMON}
                  className="next-btn"
                >
                  Next
                </button>
              </Link>
            </div>
            <div className="pokemon-description">
              <h5 className="pokemon-text info-text">Description</h5>
              {isLoadingSpecies ? (
                <p>Loading...</p>
              ) : (
                <p>
                  {
                    speciesData?.flavor_text_entries
                      .slice()
                      .reverse()
                      .find(
                        (flavor: FlavorText) => flavor?.language?.name === 'en'
                      )?.flavor_text
                  }
                </p>
              )}
            </div>
            <div className="pokemon-dimensions">
              <div className="pokemon-height">
                <h5>Height</h5>
                <span>{pokemonDetails.height / 10}m</span>
              </div>
              <div className="pokemon-weight">
                <h5>Weight</h5>
                <span>{pokemonDetails.weight / 10}kg</span>
              </div>
            </div>
            <div className="pokemon-stats">
              <h5 className="pokemon-text info-text">Stats</h5>
              <div className="parameter-container">
                {pokemonDetails?.stats.map((stat: PokemonStat) => {
                  return (
                    <div key={stat.stat.name} className="parameter-section">
                      <h5 className="info-text">{stat.stat.name}</h5>
                      <div className="statbar-container">
                        <div
                          className="statbar-segment"
                          style={{
                            backgroundColor: STAT_COLORS[stat.stat.name],
                            width: `${(stat.base_stat / 255) * 100}%`,
                          }}
                        >
                          <span>{stat.base_stat}</span>
                        </div>
                      </div>
                    </div>
                  );
                })}
              </div>
            </div>
          </>
        )}
      </div>
    </>
  );
}
