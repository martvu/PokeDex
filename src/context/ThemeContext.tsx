// ThemeContext.tsx
import { createContext, useState, ReactNode } from 'react';

interface ThemeContextType {
  darkMode: boolean;
  toggleDarkMode: () => void;
};

export const ThemeContext = createContext<ThemeContextType | undefined>(
  undefined
);

export function ThemeProvider({ children }: { children: ReactNode }) {
  const [darkMode, setDarkMode] = useState(
    localStorage.getItem('darkMode') === 'true'
  );

  const toggleDarkMode = () => {
    setDarkMode(prevDarkMode => !prevDarkMode);
    localStorage.setItem('darkMode', JSON.stringify(!darkMode));
  };

  return (
    <ThemeContext.Provider value={{ darkMode, toggleDarkMode }}>
      {children}
    </ThemeContext.Provider>
  );
}
