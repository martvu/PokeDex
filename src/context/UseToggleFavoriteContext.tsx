import { Pokemon } from '../utils/pokeApiTypes.tsx';
import { useContext, useEffect } from 'react';
import { FavoriteContext } from '../pages/PokemonList.tsx';

export function useToggleFavoriteContext(pokemonDetails: Pokemon) {
  const context = useContext(FavoriteContext);

  if (!context) {
    throw new Error(
      'useToggleFavoriteContext must be used within a FavoriteContextProvider'
    );
  }
  const { favoritesArray, setFavoritesArray } = context;

  // handle favoritesArray
  const onToggleFavorite = (pokemonId: number) => {
    if (favoritesArray.includes(pokemonId)) {
      setFavoritesArray(
        favoritesArray.filter((id: number) => id !== pokemonId)
      );
    } else {
      setFavoritesArray([...favoritesArray, pokemonId]);
    }
  };

  // save favoritesArray
  useEffect(() => {
    localStorage.setItem('favoritesArray', JSON.stringify(favoritesArray));
  }, [favoritesArray]);

  return {
    toggleFavorite: (pokemonId: number) => {
      onToggleFavorite(pokemonId); // Call the onToggleFavorite function from the context
    },
    isFavorite: favoritesArray.includes(pokemonDetails.id), // Check if the current Pokemon is a favorite
    allFavorites: favoritesArray, // Return the favoritesArray
  };
}
