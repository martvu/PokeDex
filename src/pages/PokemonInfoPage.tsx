import { usePokemonData, useSpeciesData } from '../utils/pokeApi.tsx';
import { useState } from 'react';
import { Link, useParams } from 'react-router-dom';
import {
  TYPE_COLORS,
  STAT_COLORS,
  MAX_NO_OF_POKEMON,
} from '../utils/constants.tsx';
import {
  FlavorText,
  PokemonStat,
  PokemonType,
} from '../utils/pokeApiTypes.tsx';
import { formatPokemonName, getTypeColorGradient } from '../utils/utils.tsx';
import Header from '../components/Header.tsx';
import { FavoriteIcon } from '../components/FavoriteIcon.tsx';
import { FavoriteContext } from './PokemonList.tsx';
import { HiMiniHome } from 'react-icons/hi2';

export default function PokemonInfoPage() {
  const { id } = useParams();
  const [currentId, setCurrentId] = useState<number>(id ? parseInt(id) : 1);
  const {
    data: pokemonDetails,
    isLoading,
    error,
  } = usePokemonData(currentId.toString());
  const {
    data: speciesData,
    isLoading: isLoadingSpecies,
    error: speciesError,
  } = useSpeciesData(currentId.toString());
  const [favoritesArray, setFavoritesArray] = useState<number[]>(
    JSON.parse(localStorage.getItem('favoritesArray') || '[]')
  );

  let gradient = 'black';
  if (pokemonDetails) {
    gradient = getTypeColorGradient(pokemonDetails);
  }

  return (
    <FavoriteContext.Provider value={{ favoritesArray, setFavoritesArray }}>
      <Header />
      <div className="filler-div">
        <Link to="/">
          <button className="link-home-btn">
            <HiMiniHome size={20} />
          </button>
        </Link>
      </div>
      <div className="info-page-container">
        {isLoading || (isLoadingSpecies && <div>Loading...</div>)}
        {error || (speciesError && <div>Error fetching data </div>)}
        {pokemonDetails && speciesData && (
          <>
            <div className="info-pokemon-card" style={{ background: gradient }}>
              <FavoriteIcon pokemonDetails={pokemonDetails} />
              <div className="pokemon-name-id">
                <h1>{formatPokemonName(pokemonDetails?.name)}</h1>
                <h4 className="pokemon-text">
                  {'#' + ('00' + pokemonDetails?.id).slice(-3)}
                </h4>
              </div>
              <div className="info-pokemon-img">
                <img
                  className="info-pokemon-sprite"
                  src={
                    pokemonDetails?.sprites?.other?.['official-artwork']
                      ?.front_default || '#'
                  }
                  alt={pokemonDetails.name}
                />
              </div>
              <div className="info-pokemon-data">
                <div className="info-type-list">
                  {pokemonDetails.types.map((type: PokemonType) => (
                    <span
                      key={type.slot}
                      className="info-type-badge"
                      style={{
                        backgroundColor: `${TYPE_COLORS[type.type.name]}`,
                      }}
                    >
                      {type.type.name.charAt(0).toUpperCase() +
                        type.type.name.slice(1)}
                    </span>
                  ))}
                </div>
                <div className="pokemon-genus">
                  {speciesData?.genera[7]?.genus}
                </div>
              </div>
            </div>
            <div>
              <Link to={`/pokemon/${(currentId - 1).toString()}`}>
                <button
                  onClick={() => setCurrentId(currentId - 1)}
                  disabled={currentId === 1}
                  className="prev-btn"
                >
                  Prev
                </button>
              </Link>
              <Link to={`/pokemon/${(currentId + 1).toString()}`}>
                <button
                  onClick={() => setCurrentId(currentId + 1)}
                  disabled={currentId === MAX_NO_OF_POKEMON}
                  className="next-btn"
                >
                  Next
                </button>
              </Link>
            </div>
            <div className="pokemon-description">
              <h5 className="pokemon-text info-text">Description</h5>
              <p>
                {
                  speciesData?.flavor_text_entries
                    .slice()
                    .reverse()
                    .find(
                      (flavor: FlavorText) => flavor?.language?.name === 'en'
                    )?.flavor_text
                }
              </p>
            </div>
            <div className="pokemon-dimensions">
              <div className="pokemon-height">
                <h5>Height</h5>
                <span>{pokemonDetails.height / 10}m</span>
              </div>
              <div className="pokemon-weight">
                <h5>Weight</h5>
                <span>{pokemonDetails.weight / 10}kg</span>
              </div>
            </div>
            <div className="pokemon-stats">
              <h5 className="pokemon-text info-text">Stats</h5>
              <div className="parameter-container">
                {pokemonDetails?.stats.map((stat: PokemonStat) => {
                  return (
                    <div key={stat.stat.name} className="parameter-section">
                      <h5 className="info-text">{stat.stat.name}</h5>
                      <div className="statbar-container">
                        <div
                          className="statbar-segment"
                          style={{
                            backgroundColor: STAT_COLORS[stat.stat.name],
                            width: `${(stat.base_stat / 255) * 100}%`,
                          }}
                        >
                          <span>{stat.base_stat}</span>
                        </div>
                      </div>
                    </div>
                  );
                })}
              </div>
            </div>
          </>
        )}
      </div>
    </FavoriteContext.Provider>
  );
}
