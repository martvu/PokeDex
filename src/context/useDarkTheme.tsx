import { useContext } from 'react';
import { ThemeContext } from './ThemeContext';

function useDarkTheme() {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}

export default useDarkTheme;
