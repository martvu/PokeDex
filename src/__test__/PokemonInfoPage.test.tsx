import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import PokemonInfoPage from '../pages/PokemonInfoPage';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from '../context/ThemeContext';
const queryClient = new QueryClient();

const renderPokemonInfoPage = () => {
  return render(
    <QueryClientProvider client={queryClient}>
      <ThemeProvider>
        <MemoryRouter initialEntries={['/project1']}>
          <PokemonInfoPage />
        </MemoryRouter>
      </ThemeProvider>
    </QueryClientProvider>
  );
}

describe('PokemonInfoPage', () => {
  test('pokemonInfoPage renders', async () => {
    renderPokemonInfoPage();

    expect(await screen.findByText(/#001/i)).toBeInTheDocument();
    expect(await screen.findByText(/Seed PokÃ©mon/i)).toBeInTheDocument();
    expect(await screen.findByText(/Grass/i)).toBeInTheDocument();
    expect(await screen.findByText(/Description/i)).toBeInTheDocument();

  });

  test('snapshot of pokemonInfoPage', async () => {
    const { asFragment } = renderPokemonInfoPage()
    expect(asFragment()).toMatchSnapshot();
  });
});
