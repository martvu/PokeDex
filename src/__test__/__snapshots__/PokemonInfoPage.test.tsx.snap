// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot PokemonInfoPage > snapshot of pokemonInfoPage 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="pokeball-header-icon"
    >
      <img
        alt="pokeball icon"
        src="/project1/src/assets/img/pb-icon.svg"
      />
      <div
        class="header-link-container"
      >
        <a
          class="link"
          href="/project1"
        >
          <h1>
            Pokedex
          </h1>
        </a>
      </div>
      <span
        class="switch-container"
      >
        <button
          class="theme-switch "
        >
          Turn Dark
        </button>
      </span>
    </div>
  </header>
  <div
    class="filler-div"
  />
  <div
    class="info-page-container"
  >
    <div
      class="info-home-btn"
    />
    <div
      class="info-pokemon-card"
    >
      <div
        class="pokemon-name-id"
      >
        <h1>
          Bulbasaur
        </h1>
        <h4
          class="pokemon-text"
        >
          #001
        </h4>
      </div>
      <div
        class="info-pokemon-img"
      >
        <img
          alt="bulbasaur"
          class="info-pokemon-sprite"
          src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png"
        />
      </div>
      <div
        class="info-pokemon-data"
      >
        <div
          class="info-type-list"
        >
          <span
            class="info-type-badge"
            style="background-color: rgb(122, 199, 76);"
          >
            Grass
          </span>
          <span
            class="info-type-badge"
            style="background-color: rgb(163, 62, 161);"
          >
            Poison
          </span>
        </div>
        <div
          class="pokemon-genus"
        >
          Seed Pokémon
        </div>
      </div>
    </div>
    <div>
      <a
        href="/project1/pokemon/0"
      >
        <button
          class="prev-btn"
          disabled=""
        >
          Prev
        </button>
      </a>
      <a
        href="/project1/pokemon/2"
      >
        <button
          class="next-btn"
        >
          Next
        </button>
      </a>
    </div>
    <div
      class="pokemon-description"
    >
      <h5
        class="pokemon-text info-text"
      >
        Description
      </h5>
      <p>
        Bulbasaur can be seen napping in bright sunlight.
There is a seed on its back. By soaking up the sun’s
rays, the seed grows progressively larger.
      </p>
    </div>
    <div
      class="pokemon-dimensions"
    >
      <div
        class="pokemon-height"
      >
        <h5>
          Height
        </h5>
        <span>
          NaNm
        </span>
      </div>
      <div
        class="pokemon-weight"
      >
        <h5>
          Weight
        </h5>
        <span>
          NaNkg
        </span>
      </div>
    </div>
    <div
      class="pokemon-stats"
    >
      <h5
        class="pokemon-text info-text"
      >
        Stats
      </h5>
      <div
        class="parameter-container"
      >
        <div
          class="parameter-section"
        >
          <h5
            class="info-text"
          >
            hp
          </h5>
          <div
            class="statbar-container"
          >
            <div
              class="statbar-segment"
              style="background-color: rgb(255, 1, 0); width: 17.647058823529413%;"
            >
              <span>
                45
              </span>
            </div>
          </div>
        </div>
        <div
          class="parameter-section"
        >
          <h5
            class="info-text"
          >
            attack
          </h5>
          <div
            class="statbar-container"
          >
            <div
              class="statbar-segment"
              style="background-color: rgb(240, 128, 48); width: 19.215686274509807%;"
            >
              <span>
                49
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
